services/net/filemonitor/FilemonitorAction.cs

```diff
@@ -415,35 +562,67 @@ private List<SourceContent> GetFmsArticles(string dir, IIngestServiceActionManag
+    /// dictionary will be empty and the code for the ingester will be returned. If sources is not empty it will contain a list
+    /// of paper names and their corresponding codes. In this case the code for the item is retrieved using the paperName
+    /// parameter as the key to the dictionary.
     /// </summary>
     /// <param name="ingest"></param>
     /// <param name="paperName"></param>
+    /// <param name="sources"></param>
     /// <returns></returns>
     /// <exception cref="InvalidOperationException"></exception>
-    private string GetItemSourceCode(IngestModel ingest, string paperName)
+    private string GetItemSourceCode(IngestModel ingest, string paperName, Dictionary<string, string> sources)
     {
-        // The default for selfPublished is false, so all self-published papers must have a value in configuration string.
-        var selfPublished = ingest.GetConfigurationValue<bool>(Fields.SelfPublished);
-
-        if (selfPublished)
+        if (sources.Count == 0) // Self published
         {
             return ingest.Source?.Code ?? throw new InvalidOperationException($"Ingest '{ingest.Name}' is missing source code.");
         }
         else
         {
-            // Removed the api endpoint GetIngestByNameAsync, so defaulting to the parent code for now.
-            // This is so provide working code until I can implement the "sources" attribute of the Connection string.
-            // TODO: Incomplete, this should be extracting from configuration.
-            return ingest.Source?.Code ?? throw new InvalidOperationException($"Ingest '{ingest.Name}' is missing source code.");
+            var code = "";
+            if (sources.TryGetValue(paperName, out code))
+            {
+                return code;
+            }
+            else
+            {
+                _logger.LogWarning($"Paper '{paperName}' not in configuration string for ingester '{ingest.Name}'.");
+                return "";
+            }
         }
     }
 
+    /// <summary>
+    /// Get a dictionary containing key/value pairs of the form "papername=code" extracted from the ingest record's
+    /// sources attribute.
+    /// </summary>
+    /// <param name="ingest"></param>
+    /// <returns></returns>
+    /// <exception cref="ArgumentException"></exception>
+    private Dictionary<string, string> GetSourcesForIngester(IngestModel ingest)
+    {
+        var sources = new Dictionary<string, string>();
+        var sourcesStr = ingest.GetConfigurationValue(Fields.Sources);
+        var sourcesArr = sourcesStr.Split("&");
+
+        foreach (var source in sourcesArr)
+        {
+            var pair = source.Split('=');
+
+            if (pair.Length == 2)
+            {
+                sources.Add(pair[0], pair[1]);
+            }
+            else
+            {
+                throw new ArgumentException($"Invalid source value in ingest configuration. Source '{source}' for ingest '{ingest.Name}'");
```

Comment:
Probably should be a `ConfigurationException` like the other similar errors.

Reply:
Fixed.